File: src/packages/sdk/src/runtime/execution-mode.ts
/**
 * COMPAT-204: Dual-path execution toggle
 *
 * Environment flag EVALAI_RUNTIME=legacy|spec|auto
 * Auto uses spec runtime if manifest/specs exist, else legacy
 * Existing projects continue unchanged; new projects can use DSL only
 */

import * as fs from "node:fs/promises";
import * as path from "node:path";

/**
 * Execution mode types
 */
export type ExecutionMode = "legacy" | "spec" | "auto";

/**
 * Execution mode configuration
 */
export interface ExecutionModeConfig {
  /** Current execution mode */
  mode: ExecutionMode;
  /** Whether spec runtime is available */
  hasSpecRuntime: boolean;
  /** Whether legacy runtime is available */
  hasLegacyRuntime: boolean;
  /** Project root path */
  projectRoot: string;
  /** Detected spec files */
  specFiles: string[];
  /** Detected legacy config */
  legacyConfig?: string;
}

/**
 * Get execution mode from environment or auto-detection
 */
export async function getExecutionMode(
  projectRoot: string = process.cwd(),
): Promise<ExecutionModeConfig> {
  // Check environment variable first
  const envMode = process.env.EVALAI_RUNTIME?.toLowerCase();

  if (envMode === "legacy" || envMode === "spec" || envMode === "auto") {
    return {
      mode: envMode as ExecutionMode,
      hasSpecRuntime: envMode !== "legacy",
      hasLegacyRuntime: envMode !== "spec",
      projectRoot,
      specFiles: envMode !== "legacy" ? await findSpecFiles(projectRoot) : [],
      legacyConfig: envMode !== "spec" ? await findLegacyConfig(projectRoot) : undefined,
    };
  }

  // Auto-detect mode
  return await autoDetectExecutionMode(projectRoot);
}

/**
 * Auto-detect execution mode based on project structure
 */
async function autoDetectExecutionMode(projectRoot: string): Promise<ExecutionModeConfig> {
  const specFiles = await findSpecFiles(projectRoot);
  const legacyConfig = await findLegacyConfig(projectRoot);

  const hasSpecRuntime = specFiles.length > 0;
  const hasLegacyRuntime = !!legacyConfig;

  let mode: ExecutionMode = "auto";

  // If both are available, prefer spec runtime for new projects
  if (hasSpecRuntime && hasLegacyRuntime) {
    mode = "spec"; // Prefer spec for mixed projects
  } else if (hasSpecRuntime) {
    mode = "spec";
  } else if (hasLegacyRuntime) {
    mode = "legacy";
  } else {
    mode = "auto"; // Default to auto for empty projects
  }

  return {
    mode,
    hasSpecRuntime,
    hasLegacyRuntime,
    projectRoot,
    specFiles,
    legacyConfig,
  };
}

/**
 * Find spec files in project
 */
async function findSpecFiles(projectRoot: string): Promise<string[]> {
  const specPatterns = [
    "eval/**/*.spec.ts",
    "eval/**/*.spec.js",
    "src/**/*.spec.ts",
    "src/**/*.spec.js",
    "tests/**/*.spec.ts",
    "tests/**/*.spec.js",
    "spec/**/*.ts",
    "spec/**/*.js",
  ];

  const foundFiles: string[] = [];

  for (const pattern of specPatterns) {
    try {
      const files = await searchFiles(projectRoot, pattern, projectRoot);
      foundFiles.push(...files);
    } catch (error) {
      // Ignore errors for non-existent paths
    }
  }

  // Filter for files that contain defineEval calls
  const specFilesWithDefineEval: string[] = [];

  for (const file of foundFiles) {
    try {
      const content = await fs.readFile(file, "utf-8");
      if (content.includes("defineEval")) {
        specFilesWithDefineEval.push(file);
      }
    } catch (error) {
      // Ignore read errors
    }
  }

  return specFilesWithDefineEval;
}

/**
 * Simple file search (placeholder for proper glob implementation)
 */
async function searchFiles(dir: string, pattern: string, projectRoot: string): Promise<string[]> {
  const results: string[] = [];

  try {
    const entries = await fs.readdir(dir, { withFileTypes: true });

    for (let i = 0; i < entries.length; i++) {
      const entry = entries[i];
      const fullPath = path.join(dir, entry.name);

      if (entry.isDirectory() && !entry.name.startsWith(".")) {
        results.push(...(await searchFiles(fullPath, pattern, projectRoot)));
      } else if (entry.isFile()) {
        // Simple pattern matching
        if (matchesPattern(fullPath, pattern, projectRoot)) {
          results.push(fullPath);
        }
      }
    }
  } catch (error) {
    // Ignore permission errors
  }

  return results;
}

/**
 * Simple pattern matching (placeholder for proper glob)
 */
function matchesPattern(filePath: string, pattern: string, projectRoot: string): boolean {
  const fileName = path.basename(filePath);
  const ext = path.extname(filePath);
  const dir = path.dirname(filePath);

  // Convert glob pattern to regex
  // Handle **/ and * patterns correctly
  let regexPattern = pattern;

  // Replace **/ with (?:.*/)? to match optional directory path
  regexPattern = regexPattern.replace(/\*\*\//g, "(?:.*/)?");

  // Replace remaining * with [^/]* (filename pattern)
  regexPattern = regexPattern.replace(/\*/g, "[^/]*");

  const relativePath = path.relative(projectRoot, filePath).replace(/\\/g, "/");
  const regex = new RegExp(`^${regexPattern}$`);

  return regex.test(relativePath);
}

/**
 * Find legacy config file
 */
async function findLegacyConfig(projectRoot: string): Promise<string | undefined> {
  const configPaths = [
    "evalai.config.json",
    "evalai.config.js",
    "evalai.config.ts",
    ".evalairc",
    ".evalairc.json",
  ];

  for (const configPath of configPaths) {
    const fullPath = path.join(projectRoot, configPath);
    try {
      await fs.access(fullPath);
      return fullPath;
    } catch (error) {
      // File doesn't exist, continue
    }
  }

  return undefined;
}

/**
 * Check if project can run in spec mode
 */
export function canRunSpecMode(config: ExecutionModeConfig): boolean {
  return config.hasSpecRuntime && config.specFiles.length > 0;
}

/**
 * Check if project can run in legacy mode
 */
export function canRunLegacyMode(config: ExecutionModeConfig): boolean {
  return config.hasLegacyRuntime && !!config.legacyConfig;
}

/**
 * Get recommended execution mode for project
 */
export function getRecommendedExecutionMode(config: ExecutionModeConfig): ExecutionMode {
  if (config.mode !== "auto") {
    return config.mode;
  }

  if (canRunSpecMode(config) && canRunLegacyMode(config)) {
    return "spec"; // Prefer spec for mixed projects
  }

  if (canRunSpecMode(config)) {
    return "spec";
  }

  if (canRunLegacyMode(config)) {
    return "legacy";
  }

  return "auto";
}

/**
 * Validate execution mode compatibility
 */
export function validateExecutionMode(config: ExecutionModeConfig): {
  valid: boolean;
  warnings: string[];
  errors: string[];
} {
  const warnings: string[] = [];
  const errors: string[] = [];

  // Check for mixed project
  if (config.hasSpecRuntime && config.hasLegacyRuntime) {
    warnings.push(
      "Project contains both spec files and legacy config. " +
        "Consider migrating legacy tests to spec format for consistency.",
    );
  }

  // Check for no runtime
  if (!config.hasSpecRuntime && !config.hasLegacyRuntime) {
    warnings.push(
      "No tests found. This appears to be an empty project. " +
        "Use 'evalai init' to create a new project.",
    );
  }

  // Check for spec mode without spec files
  if (config.mode === "spec" && !canRunSpecMode(config)) {
    errors.push(
      "Spec mode requested but no spec files found. " +
        "Create spec files with defineEval() or use legacy mode.",
    );
  }

  // Check for legacy mode without config
  if (config.mode === "legacy" && !canRunLegacyMode(config)) {
    errors.push(
      "Legacy mode requested but no evalai.config.json found. " +
        "Create a config file or use spec mode.",
    );
  }

  return {
    valid: errors.length === 0,
    warnings,
    errors,
  };
}

/**
 * Print execution mode information
 */
export function printExecutionModeInfo(config: ExecutionModeConfig): void {
  console.log(`🔧 EvalAI Execution Mode: ${config.mode.toUpperCase()}`);
  console.log(`📁 Project root: ${config.projectRoot}`);
  console.log(``);

  if (config.hasSpecRuntime) {
    console.log(`✅ Spec runtime available`);
    console.log(`   Found ${config.specFiles.length} spec file(s):`);
    config.specFiles.slice(0, 5).forEach((file) => {
      console.log(`   - ${path.relative(config.projectRoot, file)}`);
    });
    if (config.specFiles.length > 5) {
      console.log(`   ... and ${config.specFiles.length - 5} more`);
    }
  } else {
    console.log(`❌ No spec runtime found`);
  }

  console.log(``);

  if (config.hasLegacyRuntime) {
    console.log(`✅ Legacy runtime available`);
    if (config.legacyConfig) {
      console.log(`   Config: ${path.relative(config.projectRoot, config.legacyConfig)}`);
    }
  } else {
    console.log(`❌ No legacy runtime found`);
  }

  console.log(``);

  const validation = validateExecutionMode(config);

  if (validation.warnings.length > 0) {
    console.log(`⚠️  Warnings:`);
    validation.warnings.forEach((warning) => console.log(`   ${warning}`));
    console.log(``);
  }

  if (validation.errors.length > 0) {
    console.log(`❌ Errors:`);
    validation.errors.forEach((error) => console.log(`   ${error}`));
    console.log(``);
  }

  const recommended = getRecommendedExecutionMode(config);
  console.log(`💡 Recommended mode: ${recommended.toUpperCase()}`);

  if (config.mode === "auto") {
    console.log(`🔄 Auto mode will use: ${recommended.toUpperCase()}`);
  }
}

/**
 * Environment variable helpers
 */
export const ENV_VARS = {
  EXECUTION_MODE: "EVALAI_RUNTIME",
  POSSIBLE_VALUES: ["legacy", "spec", "auto"],
  DEFAULT: "auto",
} as const;

/**
 * Check if environment variable is set
 */
export function hasExecutionModeEnv(): boolean {
  return !!process.env.EVALAI_RUNTIME;
}

/**
 * Get current environment variable value
 */
export function getExecutionModeEnv(): string | undefined {
  return process.env.EVALAI_RUNTIME;
}

/**
 * Set execution mode environment variable
 */
export function setExecutionModeEnv(mode: ExecutionMode): void {
  process.env.EVALAI_RUNTIME = mode;
}

/**
 * Clear execution mode environment variable
 */
export function clearExecutionModeEnv(): void {
  delete process.env.EVALAI_RUNTIME;
}

File: src/packages/sdk-python/src/evalgate_sdk/runtime/execution_mode.py
"""Dual-path execution mode detection (T8).

Port of the TypeScript SDK's ``execution-mode.ts``.
Environment flag ``EVALGATE_RUNTIME=legacy|spec|auto``.
"""

from __future__ import annotations

import os
from dataclasses import dataclass, field
from pathlib import Path
from typing import Literal

ExecutionMode = Literal["legacy", "spec", "auto"]

ENV_VAR = "EVALGATE_RUNTIME"
POSSIBLE_VALUES = ("legacy", "spec", "auto")
DEFAULT_MODE: ExecutionMode = "auto"


@dataclass
class ExecutionModeConfig:
    mode: ExecutionMode = "auto"
    has_spec_runtime: bool = False
    has_legacy_runtime: bool = False
    project_root: str = ""
    spec_files: list[str] = field(default_factory=list)
    legacy_config: str | None = None


def find_spec_files(project_root: str) -> list[str]:
    """Search for Python files containing ``define_eval`` calls."""
    root = Path(project_root)
    patterns = [
        "eval/**/*.py",
        "evals/**/*.py",
        "src/**/*.py",
        "tests/**/*.py",
        "spec/**/*.py",
    ]

    found: list[str] = []
    for pattern in patterns:
        for f in root.glob(pattern):
            if f.name.startswith("_"):
                continue
            try:
                content = f.read_text(encoding="utf-8", errors="ignore")
                if "define_eval" in content:
                    found.append(str(f))
            except OSError:
                continue
    return found


def find_legacy_config(project_root: str) -> str | None:
    """Search for legacy config files."""
    root = Path(project_root)
    candidates = [
        "evalai.config.json",
        "evalai.config.py",
        "evalgate.config.json",
        ".evalgaterc",
        ".evalgaterc.json",
        ".evalgate/config.json",
        ".evalai/config.json",
    ]
    for name in candidates:
        p = root / name
        if p.exists():
            return str(p)
    return None


def get_execution_mode(project_root: str | None = None) -> ExecutionModeConfig:
    """Determine execution mode from environment or auto-detection."""
    root = project_root or os.getcwd()
    env_mode = os.environ.get(ENV_VAR, "").lower()

    if env_mode in POSSIBLE_VALUES:
        spec_files = find_spec_files(root) if env_mode != "legacy" else []
        legacy_config = find_legacy_config(root) if env_mode != "spec" else None
        return ExecutionModeConfig(
            mode=env_mode,  # type: ignore[arg-type]
            has_spec_runtime=env_mode != "legacy",
            has_legacy_runtime=env_mode != "spec",
            project_root=root,
            spec_files=spec_files,
            legacy_config=legacy_config,
        )

    # Auto-detect
    spec_files = find_spec_files(root)
    legacy_config = find_legacy_config(root)
    has_spec = len(spec_files) > 0
    has_legacy = legacy_config is not None

    if has_spec:
        mode: ExecutionMode = "spec"
    elif has_legacy:
        mode = "legacy"
    else:
        mode = "auto"

    return ExecutionModeConfig(
        mode=mode,
        has_spec_runtime=has_spec,
        has_legacy_runtime=has_legacy,
        project_root=root,
        spec_files=spec_files,
        legacy_config=legacy_config,
    )


def can_run_spec_mode(config: ExecutionModeConfig) -> bool:
    return config.has_spec_runtime and len(config.spec_files) > 0


def can_run_legacy_mode(config: ExecutionModeConfig) -> bool:
    return config.has_legacy_runtime and config.legacy_config is not None


def get_recommended_mode(config: ExecutionModeConfig) -> ExecutionMode:
    if config.mode != "auto":
        return config.mode
    if can_run_spec_mode(config):
        return "spec"
    if can_run_legacy_mode(config):
        return "legacy"
    return "auto"


def validate_execution_mode(config: ExecutionModeConfig) -> dict[str, list[str] | bool]:
    """Validate execution mode compatibility."""
    warnings: list[str] = []
    errors: list[str] = []

    if config.has_spec_runtime and config.has_legacy_runtime:
        warnings.append(
            "Project contains both spec files and legacy config. Consider migrating legacy tests to spec format."
        )

    if not config.has_spec_runtime and not config.has_legacy_runtime:
        warnings.append("No tests found. Use 'evalgate init' to create a new project.")

    if config.mode == "spec" and not can_run_spec_mode(config):
        errors.append(
            "Spec mode requested but no spec files found. Create spec files with define_eval() or use legacy mode."
        )

    if config.mode == "legacy" and not can_run_legacy_mode(config):
        errors.append("Legacy mode requested but no config file found. Create a config file or use spec mode.")

    return {"valid": len(errors) == 0, "warnings": warnings, "errors": errors}


def has_execution_mode_env() -> bool:
    return ENV_VAR in os.environ


def get_execution_mode_env() -> str | None:
    return os.environ.get(ENV_VAR)


def set_execution_mode_env(mode: ExecutionMode) -> None:
    os.environ[ENV_VAR] = mode


def clear_execution_mode_env() -> None:
    os.environ.pop(ENV_VAR, None)
